src/lib/poptrie_lookup.dasl

```diff
@@ -0,0 +1,158 @@
+|.define dmap,    rcx -- pointer to directmap
+|.define index,   r8d -- index into node array
+|.define node,    r8  -- pointer into node array 
+|.define offset,  r9  -- offset into key
+|.define v,       r10 -- k or s bits extracted from key
+|.define vec,     r11 -- 64-bit vector or leafvec
+
+-- lookup(leaf_t *leaves, node_t *nodes, key) -> leaf_t
+function lookup (Dst, Poptrie, keysize)
+   if Poptrie.direct_pointing then
+      -- v = extract(key, 0, Poptrie.s)
+      local direct_mask = bit.lshift(1ULL, Poptrie.s) - 1
+      -- v = band(key, direct_mask)
+      | mov v, direct_mask
+      | and v, key
+      -- index = dmap[v]
+      | mov index, dword [dmap+v*4]
+      -- eax = band(index, leaf_tag)
+      | mov eax, 0x80000000
+      | and eax, index
+      -- is leaf_tag set?
+      | cmp eax, 0
+      | je >1 -- leaf_tag not set, index is a node
+      -- eax = leaf_tag - 1 (tag inverted)
+      | mov eax, 0x7FFFFFFF
+      | and eax, index
+      | ret
+      -- node, offset = nodes[index], s
+      |1:
+      | imul index, 24 -- multiply by node size
+      | lea node, [nodes+index]
+      -- offset = s
+      | mov offset, 18
+   else
+      -- index, node, offset = 0, nodes[index], 0
+      | mov index, 0
+      | lea node, [nodes+0] -- nodes[0]
+      | mov offset, 0
+   end
+   -- while band(vec, lshift(1ULL, v)) ~= 0
+   |2:
+   -- v = extract(key, offset, k=6)
+   if keysize <= 64 then
+      -- v = rshift(key, offset)
+      | mov v, key
+      | mov rcx, offset
+      | shr v, cl
+      -- v = band(v, lshift(1, k=6) - 1)
+      | and v, 0x3F
+   else error("NYI") end
+   -- vec = nodes[index].vector
+   | mov vec, qword [node+8]
+   -- rax = lshift(1ULL, v)
+   | mov rax, 1
+   | mov rcx, v
+   | shl rax, cl
+   -- rax = band(vec, rax)
+   | and rax, vec
+   -- is bit v set in vec?
+   | cmp rax, 0
+   | je >3 -- reached leaf, exit loop
+   -- rax = lshift(2ULL, v) - 1
+   | mov rax, 2
+   | mov rcx, v
+   | shl rax, cl
+   | sub rax, 1
+   -- rax = popcnt(band(vec, rax))
+   | and rax, vec
+   | popcnt rax, rax
+   -- index = base + bc - 1
+   | mov index, dword [node+20] -- nodes[index].base1
+   | add index, eax
+   | sub index, 1
+   -- node = nodes[index]
+   | imul index, 24 -- multiply by node size
+   | lea node, [nodes+index]
+   -- offset = offset + k
+   | add offset, 6
+   | jmp <2 -- loop
+   -- end while
+   |3:
+   -- rax = lshift(2ULL, v) - 1
+   | mov rax, 2
+   | mov rcx, v
+   | shl rax, cl
+   | sub rax, 1
+   if Poptrie.leaf_compression then
+      -- vec = nodes[index].leafvec
+      | mov vec, qword [node+0]
+   else error("NYI") end
+   -- rax = popcnt(band(vec, rax)) - 1
+   | and rax, vec
+   | popcnt rax, rax
+   -- return leaves[base + bc - 1]
+   | mov index, dword [node+16] -- nodes[index].base0
+   | add index, eax
+   | sub index, 1
+   -- assumption: at most ax is set at this point because popcnt(qword) < byte
+   | mov ax, word [leaves+index*2] -- leaves[index]
```

Comment:
This would probably be better as `movzx eax, ...`. The assumption in the comment is true, but the CPU is not clever enough to know this, and so if you do a `mov` into `ax`, then when `eax` is next used, it'll insert a micro-up to combine `ax` and the previous high 16-bits of `eax` (whereas `movzx` here is the same price as `mov`, and doesn't have the extra micro-op penalty down the line).